### ISSUE TITLE
Joint prediction problem

### ISSUE THREAD

**Issue by CodingNovice7** [AutodeskAILab/JoinABLe#4]
Hello, this project is really cool!
I am trying to reproduce your results, but I am having some problems understanding the joint axis prediction section through the paper and the code, and would appreciate your help.
The specific questions are as follows: In the EdgeMLPMPN part of the model outputs a vector of size G1 nodes * G2 nodes, does this vector represent whether the nodes in G1 and the nodes in G2 are adjacent to the binaries result? And what does the adjacency here indicate? Contact? Is a softmax operation done on the output during testing to distinguish the pair of B-REP faces and edge entities chosen by the designer?
Thank you as always and have a nice day !

**Comment by karldd** [comment]
> In the EdgeMLPMPN part of the model outputs a vector of size G1 nodes * G2 nodes, does this vector represent whether the nodes in G1 and the nodes in G2 are adjacent to the binaries result? 

`EdgeMLPMPN` performs message passing in the joint graph (connecting the two bodies) to predict edge logits by using an MLP (1x1 convolutions). If you look at the figure below, the joint graph is shown as a matrix. It is a densely connected graph between two B-Rep bodies. The idea is to get unique embeddings for every combination of all connections between the two B-Rep bodies. 

![image](https://user-images.githubusercontent.com/619336/198149025-939c0912-bf8b-4aeb-95d4-c6c89b9bfa82.png)

> And what does the adjacency here indicate? Contact? 

So there isn't adjacency here, but rather dense connections between the two B-Rep bodies.

> Is a softmax operation done on the output during testing to distinguish the pair of B-REP faces and edge entities chosen by the designer?

Yes [this](https://github.com/AutodeskAILab/JoinABLe/blob/main/joint/joint_prediction_set.py#L124-L133) is what that looks like, where `x` has the logits for the densely connected joint graph (aka matrix).

```python
    @staticmethod
    def get_network_predictions(g1, g2, joint_graph, model):
        """Get the network predictions"""
        fake_batch = (
            Batch.from_data_list([g1]),
            Batch.from_data_list([g2]),
            Batch.from_data_list([joint_graph]),
        )
        x = model(fake_batch)
        prob = F.softmax(x, dim=0)
        return prob.view(g1.num_nodes, g2.num_nodes)
```

**Comment by nylvy** [comment]
Thank you for your timely reply!
But on this issue I still do not understand some.
What is an dense connection? What is the role of dense connections in this joint work? Is the joint axis prediction task to predict the two interconnected B-REP entities selected by the designer for assembly? Are the two joint axes aligned by determining whether the angular deviation of the joint axes of these two B-REP entities is within the threshold (10° as stated in the paper)?
Thank you as always and have a nice day !

**Comment by CodingNovice7** [comment]
Thank you for your timely reply!
But on this issue I still do not understand some.
What is an dense connection? What is the role of dense connections in this joint work? Is the joint axis prediction task to predict the two interconnected B-REP entities selected by the designer for assembly? Are the two joint axes aligned by determining whether the angular deviation of the joint axes of these two B-REP entities is within the threshold (10° as stated in the paper)?
Thank you as always and have a nice day !

**Comment by karldd** [comment]
Here is a simple visual example. We have two graphs, one for each B-Rep. I'm showing these as two node graphs for simplicity.
<img width="351" alt="image" src="https://user-images.githubusercontent.com/619336/198329005-e0b3b5a5-b906-4d39-9a5b-1bce25fe2b3e.png">

Making the dense connection graph (aka matrix) would look like this:
<img width="319" alt="image" src="https://user-images.githubusercontent.com/619336/198329205-48bac887-5501-44b1-8156-7c16942e7b60.png">

>What is the role of dense connections in this joint work?

The role is to make unique embeddings for every pair of entities across the two B-Reps.


> Is the joint axis prediction task to predict the two interconnected B-REP entities selected by the designer for assembly?

Yes exactly, and it is the unique embeddings for each pair of entities that form the prediction space.

>Are the two joint axes aligned by determining whether the angular deviation of the joint axes of these two B-REP entities is within the threshold (10° as stated in the paper)?

No, the 10 degree threshold is just for the point cloud baselines. The joint axes are determined based on the entities predicted. You can see [the code for that here](https://github.com/AutodeskAILab/JoinABLe/blob/main/joint/joint_axis.py#L14).

**Comment by CodingNovice7** [comment]
Oh, is the output of this logit equivalent to a one-dimensional embedding vector? This embedding should be useful later, right? I guess I need to go back and look at your code later on. Thanks for your patience!
Thank you as always and have a nice day !

**Comment by karldd** [comment]
The output is currently 1D logits, but you could modify the network to return the full embedding vector quite easily. So this would give you an `n x m x d` tensor, where `n` and `m` are the number of nodes in each graph and `d` is the hidden dimension size I believe.

**Comment by CodingNovice7** [comment]
In the last two days, I have debugged your open source code completely, and there are some places that I don't quite understand in combination with the code and the paper, so I hope you can guide me at your convenience.
1.Regarding the joint pose search after the joint axis prediction, you have used some method to calculate a quantity called offset_limit to limit the subsequent search range. I tried to check the link given in the code comment, but that explanation is not very friendly for people who do not have basic knowledge in this area, so if it is convenient, I hope you can give some basic explanation about this area.
2.In the return value of the best parameter, why does offset return a one-dimensional quantity, and does this one-dimensional quantity represent the compensation distance between two joint axes? 
3.In the search phase for joint poses you use the Nelder-Mead algorithm and the conditions for meeting the required accuracy do not seem to be explicitly given in the paper.
Thank you as always and have a nice day !

**Comment by karldd** [comment]
1. This is the code [here](https://github.com/AutodeskAILab/JoinABLe/blob/main/search/search_base.py#L86). The comments explain what it is doing.
2. It is 1 dimensional as it is the offset along the joint axis of each body. If you play around with a joint in Fusion 360 it might help visualize by changing the offset value and seeing what it looks like in 3D.
3. Code is [here](https://github.com/AutodeskAILab/JoinABLe/blob/main/search/search_simplex.py) for the Nelder-Mead. I'm not sure what you mean by 'required accuracy'. In the paper we report Chamfer Distance for the joint pose prediction task.

**Comment by lylwy** [comment]
Thanks for your reply, I understood the related work in the paper well through your code. I am now eager to reproduce the future application part you mentioned in your paper, i.e., the problem of multi-part assembly given an assembly sequence. What I am more curious about is how JoinABLe completes the assembly with the given sequence in a multi-part assembly task. For a multi-part assembly task, for example, a three-part assembly task, for clarity, I use A,B,C to represent these three parts here, now I am given the assembly sequence of these three parts as A=1,B=2,C=3, the first step input the graph structure of A and B to generate a component D, then the next step I input this component D and C to JoinABLe for assembly or input B and C?  If you enter B and C, how do you ensure that when B and C are assembled by JoinABLe, C interferes with A, which has already finished assembling with B? If you enter B and D, does that require the generation of many intermediate components like D for an assembly task with more parts?

**Comment by CodingNovice7** [comment]
Thanks for your reply, I understood the related work in the paper well through your code. I am now eager to reproduce the future application part you mentioned in your paper, i.e., the problem of multi-part assembly given an assembly sequence. 


What I am more curious about is how JoinABLe completes the assembly with the given sequence in a multi-part assembly task. For a multi-part assembly task, for example, a three-part assembly task, for clarity, I use A,B,C to represent these three parts here, now I am given the assembly sequence of these three parts as A=1,B=2,C=3, the first step input the graph structure of A and B to generate a component D, then the next step I input this component D and C to JoinABLe for assembly or input B and C? If you enter B and C, how do you ensure that when B and C are assembled by JoinABLe, C interferes with A, which has already finished assembling with B? If you enter B and D, does that require the generation of many intermediate components like D for an assembly task with more parts?

**Comment by yunshengtian** [comment]
For multi-part assembly with A=1, B=2, C=3, what we do is first assemble A+B->D, then assemble D+C to be the final assembly, which does not have the interference issue but indeed requires generating all the intermediate components. If you would like to see more details, please refer to appendix A.3.1 of our paper. Thanks for your interest.

**Comment by CodingNovice7** [comment]
> For multi-part assembly with A=1, B=2, C=3, what we do is first assemble A+B->D, then assemble D+C to be the final assembly, which does not have the interference issue but indeed requires generating all the intermediate components. If you would like to see more details, please refer to appendix A.3.1 of our paper. Thanks for your interest.

I understand! 
Thank you as always and have a nice day !

**Comment by ttsesm** [comment]
@lylwy did you manage to have joinable running for multi-part assembly tasks? I would be interested to this as well.

Thanks.

**Comment by CodingNovice7** [comment]
> @lylwy did you manage to have joinable running for multi-part assembly tasks? I would be interested to this as well.
> 
> Thanks.

I am currently working on.
Thanks!

**Comment by ttsesm** [comment]
If you have any update, I would be happy if you could share some progress ;-). My idea is to do something similar as well also in combination with the recent Assemble-Them-All work from the other repository. So on my spare time I am trying to work with it.

Thanks.
